clippy_lints/src/loops/infinite_loops.rs

```diff
@@ -0,0 +1,99 @@
+use rustc_lint::LateContext;
+
+use super::INFINITE_LOOPS;
+
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    loop_block: &Block<'_>,
+    label: Option<Label>,
+    parent_fn_ret_ty: FnRetTy<'_>,
+) {
+    if is_lint_allowed(cx, INFINITE_LOOPS, expr.hir_id)
+        || matches!(
+            parent_fn_ret_ty,
+            FnRetTy::Return(Ty {
+                kind: TyKind::Never,
+                ..
+            })
+        )
+    {
+        return;
+    }
+
+    // First, find any `break` or `return` without entering any inner loop,
+    // then, find `return` or labeled `break` which breaks this loop with entering inner loop,
+    // otherwise this loop is a infinite loop.
+    let mut direct_br_or_ret_finder = BreakOrRetFinder::default();
+    direct_br_or_ret_finder.visit_block(loop_block);
+
+    let is_finite_loop = direct_br_or_ret_finder.found || {
+        let mut inner_br_or_ret_finder = BreakOrRetFinder {
+            label,
+            enter_nested_loop: true,
+            ..Default::default()
+        };
+        inner_br_or_ret_finder.visit_block(loop_block);
+        inner_br_or_ret_finder.found
+    };
+
+    if !is_finite_loop {
+        span_lint_and_then(cx, INFINITE_LOOPS, expr.span, "infinite loop detected", |diag| {
+            if let FnRetTy::DefaultReturn(ret_span) = parent_fn_ret_ty {
+                diag.span_suggestion(
+                    ret_span,
+                    "if this is intentional, consider specifing `!` as function return",
+                    " -> !",
+                    Applicability::MaybeIncorrect,
+                );
+            } else {
+                diag.span_help(
+                    expr.span,
+                    "if this is not intended, add a `break` or `return` condition in this loop",
+                );
+            }
+        });
+    }
+}
+
+#[derive(Default)]
+struct BreakOrRetFinder {
+    label: Option<Label>,
+    found: bool,
+    enter_nested_loop: bool,
+}
+
+impl<'hir> Visitor<'hir> for BreakOrRetFinder {
+    fn visit_expr(&mut self, ex: &'hir Expr<'_>) {
+        match &ex.kind {
+            ExprKind::Break(Destination { label, .. }, ..) => {
+                // When entering nested loop, only by breaking this loop's label
+                // would be considered as exiting this loop.
+                if self.enter_nested_loop {
+                    if label.is_some() && *label == self.label {
+                        self.found = true;
+                    }
+                } else {
+                    self.found = true;
+                }
+            },
+            ExprKind::Ret(..) => self.found = true,
+            ExprKind::Loop(..) if !self.enter_nested_loop => (),
+            // `panic` like macro calls
+            ExprKind::Call(caller, _) => {
+                println!("caller: {:#?}", caller);
+            }
+            _ => {
+                println!("expr: {:?}, span: {:?}", ex.precedence(), ex.span);
```

Comment:
in this leftover debug stuff (might be why the tests fail)

Reply:
Ah this :rofl: , haha, I was testing the things that you suggested, and suddenly decided I want to save the work and sleep, so this happened, don't worry I'll fix it~

Reply:
Tests should be good now.

Btw, how can I optimize this? :thinking: Right now it has to run the same visitor with similar behavior twice, and might revisiting the same loop multiple times

Reply:
No ideas sorry, :sweat: 
Maybe someone in the clippy channel on zulip has an idea?

Reply:
Why does the visitor run twice?
Can't you look for a `break` anywhere, even in nested loops, in a single visitor "run"?

I think you could do that by storing an int in the visitor for the loop depth, increment/decrement when visiting/leaving one and then if the depth is 0 then `break` without a label means that it breaks the relevant loop

Reply:
> Why does the visitor run twice? Can't you look for a `break` anywhere, even in nested loops, in a single visitor "run"?
> 
> I think you could do that by storing an int in the visitor for the loop depth, increment/decrement when visiting/leaving one and then if the depth is 0 then `break` without a label means that it breaks the relevant loop

Well, I've thought about that, but unfortunately I don't know how to identify where the visitor is leaving a loop... I tried to find methods that are similar to `check_xxx_post` etc, but no luck :disappointed:... So I gave up

Reply:
>  I don't know how to identify where the visitor is leaving a loop

`check_expr_post`,  then checking `if let ExprKind::Loop = ex.kind`.
Either that, or you could do it directly within `check_expr` I guess
```rs
match ex.kind {
  // ...
  ExprKind::Loop(..) => {
    self.loop_depth += 1;
    intravisit::walk_expr(self, ex);
    self.loop_depth -= 1;
  }
}
```

Reply:
> or you could do it directly within `check_expr` I guess

Excellent! It works, thank you~